### PR TITLE
Fix issue 16004 - Document changes to protection attributes

### PR DESCRIPTION
**Pull request by mathias-lang-sociomantic**
- Rename to visibility attribute
- Expand documentation on name lookup in module doc

CC @MartinNowak

### CODE REVIEW

**Comment by mathias-lang-sociomantic** [comment]
Ping @MartinNowak

**Comment by mathias-lang-sociomantic** [comment]
Anyone up for reviewing this ?

File: spec/attribute.dd [hunk 1]
```diff
@@ -346,76 +346,51 @@ $(GNAME ProtectionAttribute):
...
+        $(P They may be refered to as protection attributes in documents written
+        before $(LINK2 http://wiki.dlang.org/DIP22, DIP22) was implemented.)
+
+        $(P Symbols with $(D private) visibility can only be accessed from
+        within the same module.
+        Private member functions are implicitly final and cannot be overridden.
         )
 
-        $(P Package extends private so that package members can be accessed
+        $(P $(D package) extends private so that package members can be accessed
```
Comment (qznc) [inline spec/attribute.dd]: Modern style is to use backticks: \`package\`

File: spec/attribute.dd [hunk 2]
```diff
@@ -346,76 +346,51 @@ $(GNAME ProtectionAttribute):
...
         from code in other modules that are in the same package.
-        This applies to the innermost package only, if a module is in
-        nested packages.
+        If no identifier is provided, this applies to the innermost package only,
+        or defaults to $(D private) if a module is not nested in a package.
         )
 
-        $(P Package may have an optional parameter - dot-separated identifier
+        $(P $(D package) may have an optional parameter - dot-separated identifier
         list which is resolved as the qualified package name. If this optional
```
Comment (qznc) [inline spec/attribute.dd]: I don't understand that minus. Why not "optional parameter, a dot-separated"?

File: spec/module.dd [hunk 3]
```diff
@@ -219,26 +219,46 @@ $(GNAME ModuleAliasIdentifier):
...
+$(H3 $(LNAME2 name_lookup, Symbol Name Lookup))
 
 	$(P The simplest form of importing is to just list the
 	modules being imported:)
 
 ---------
+module myapp.main;
+
 import std.stdio; // import module stdio from package std
-import foo, bar;  // import modules foo and bar
```
Comment (qznc) [inline spec/module.dd]: This somewhat hides the comma-separation syntax. Do we actually want to discourage that? Fine with me. :smile:

**Comment by qznc** [comment]
LGTM, except that backticks should be used these days.

**Comment by CyberShadow** [comment]
> LGTM, except that backticks should be used these days.

I would consider this an issue if some effort had been made to remove all instances of `$(D ...)` and replace it with backticks, which has not been done, thus I would not consider this an issue.

File: spec/attribute.dd [hunk 4]
```diff
@@ -346,76 +346,51 @@ $(GNAME ProtectionAttribute):
...
+        all its descendants.
         )
 
-        $(P Protected means that only members of the enclosing class or any
-        classes derived from that class,
-        or members and functions in the same module
-        as the enclosing class, can access the member.
+        $(P $(D protected) only applies inside classes (and templates as they can be mixed in)
+        and means that a symbol may only be accessed by members of the same module,
+        or by a derived class.
```
Comment (MartinNowak) [inline spec/attribute.dd]: can only be seen

File: spec/attribute.dd [hunk 5]
```diff
@@ -346,76 +346,51 @@ $(GNAME ProtectionAttribute):
...
-        that member can only be accessed for the object instance
+        function, that member can only be accessed for the object instance
         which can be implicitly cast to the same type as $(SINGLEQUOTE this).
-        Protected module members are illegal.
+        $(D protected) module members are illegal.
         )
 
-        $(P Public means that any code within the executable can access the member.
+        $(P $(D public) means that any code within the executable can access the member.
+        It is the default visibility attribute.
```
Comment (MartinNowak) [inline spec/attribute.dd]: can see

**Comment by MartinNowak** [comment]
Another pass to replace accessing with seeing and we're fine.

**Comment by mathias-lang-sociomantic** [comment]
Kinda forgot about it :blush: 

```diff
diff --git a/spec/attribute.dd b/spec/attribute.dd
index deb77d1..852f6e4 100644
--- a/spec/attribute.dd
+++ b/spec/attribute.dd
@@ -365,14 +365,14 @@ $(GNAME VisibilityAttribute):
         or defaults to $(D private) if a module is not nested in a package.
         )
 
-        $(P $(D package) may have an optional parameter - dot-separated identifier
+        $(P $(D package) may have an optional parameter, dot-separated identifier
         list which is resolved as the qualified package name. If this optional
         parameter is present, the symbol will be visible by this package and
         all its descendants.
         )
 
         $(P $(D protected) only applies inside classes (and templates as they can be mixed in)
-        and means that a symbol may only be accessed by members of the same module,
+        and means that a symbol can only be seen by members of the same module,
         or by a derived class.
         If accessing a protected instance member through a derived class member
         function, that member can only be accessed for the object instance
@@ -380,7 +380,7 @@ $(GNAME VisibilityAttribute):
         $(D protected) module members are illegal.
         )
 
-        $(P $(D public) means that any code within the executable can access the member.
+        $(P $(D public) means that any code within the executable can see the member.
         It is the default visibility attribute.
         )
 ```

**Comment by MartinNowak** [comment]
> Kinda forgot about it

Sorry, should have done it myself, but had negative free time lately.

**Pull request merged**